### ISSUE TITLE
A doubt on the right choice for ring and field type in the SDG_Linf_2 testsuite

### ISSUE THREAD

**Issue by afabri** [CGAL/cgal#731]
The file test_sdg_linf_2_et.cpp  defines `Gmpq` as ring and as field type.
When GMP is not used we use `MP_Float` instead, but define
`MP_Float` as ring type and `Quotient<MP_Float>` as field type.
This leads to a failing [testcase](https://cgal.geometryfactory.com/CGAL/Members/testsuite/CGAL-4.8-Ic-115/Segment_Delaunay_graph_Linf_2/TestReport_cgaltester_x86-64_Darwin-13.0_Apple-clang-5.1_Release-LEDA-without-GMP.gz)

Making both `Quotient<MP_Float>`  (as in the SDG_2 testsuite) compiles, but then the names seem
not to be choses correctly. 

If the types should be different the testcase shows that we have to make fixes in the predicates.

**Comment by lrineau** [comment]
`CGAL::Simple_cartesian<CGAL::MP_Float>` cannot compile, because `CGAL::MP_Float` is not a field type, but only a ring type.

The following error:
```
CGAL/constructions/kernel_ftC2.h:40:15: error: invalid operands to binary expression ('CGAL::MP_Float' and 'int')
  x = (px+qx) / 2;
      ~~~~~~~ ^ ~
```
shows that the implementation of `midpoint` for the cartesian kernel does not use our concepts. The division by the `int` 2 is not allowed. It should be `x = (px+qx) / FT(2)`.

**Comment by lrineau** [comment]
`K_ring`, the kernel using a rind type, should use the homogeneous kernel.

**Comment by lrineau** [comment]
Try this patch:
```
--- a/Segment_Delaunay_graph_Linf_2/test/Segment_Delaunay_graph_Linf_2/test_sdg_linf_2_et.cpp
+++ b/Segment_Delaunay_graph_Linf_2/test/Segment_Delaunay_graph_Linf_2/test_sdg_linf_2_et.cpp
@@ -37,11 +37,12 @@ typedef exact_ring_t   ring_number_t;
 typedef exact_field_t  field_number_t;
 
 #include <CGAL/Simple_cartesian.h>
+#include <CGAL/Simple_homogeneous.h>
 
 #include <CGAL/Segment_Delaunay_graph_Linf_2.h>
 #include <CGAL/Segment_Delaunay_graph_Linf_traits_2.h>
 
-struct K_ring  : public CGAL::Simple_cartesian<ring_number_t> {};
+struct K_ring  : public CGAL::Simple_homogeneous<ring_number_t> {};
 struct K_field : public CGAL::Simple_cartesian<field_number_t> {};
 
 typedef CGAL::Field_tag  MTag;
```

**Comment by philaris** [comment]
I tried to apply the fix that Laurent suggested but there is still a similar error. I hope I will have a look during the weekend. Because of the setup_insert_on_pointers_l2() error, I think some code in SDG L2  creates some problems only when used in the SDG Linf code (which is derived from the SDG L2) and also only when the no intersecting segments variant of the code is used.

**Comment by afabri** [comment]
I think the starting point should be to understand, if conceptually there should be a ring or field type in the testfiles.

**Comment by philaris** [comment]
As far as I remember, from Menelaos' @mkaravel code, the ring
types are used when there are non-intersecting segments.

I had a look at the issue and the first error messages.

The problem appears in the non-intersecting segments test
(CGAL::test_no_x). In this test, the SDG Linf class for
non-intersecting segments SDG2_wi is used. This class derives from
the corresponding SDG L2 class.  The base class sets up two
pointers that have to do with functions called when two segments
intersect or when a point is inserted in the interior of an
existing segment. 

This setup is done in the following function:
setup_insert_on_pointers_l2()

The code that sets up the pointer is the following:
insert_exact_point_on_segment_ptr = &Self::insert_exact_point_on_segment;

I introduced these two pointers in the following commit:
https://github.com/philaris/cgal/commit/8cf9a96b459f5ee235ebcb577adfbc106acfc5bf
in order to reuse as much code as possible from SDG L2, without
changing much the existing code.

Even with GMP, I had a problem with one of the pointers
(insert_point_on_segment_ptr) and I had to set it to NULL in the
non-intersectings segments case:
CGAL/cgal@92bda0623caf0e0e119633781c0fe25dddcb6639

I am now also setting the other pointer
(insert_exact_point_on_segment_ptr) in the non-intersecting case.
The changes are in the branch sdg-ptr-fix-philaris.

Should I try to integrate and see what happens?

Some of the errors go away, but there are some more issues.
I will look what can be done with the rest of the errors.

**Comment by sgiraudot** [comment]
@philaris Have you made any progress with these errors?

**Comment by philaris** [comment]
@sgiraudot Sorry for the late reply. Unfortunately, I have been busy and not able to make any progress since the last commit in the above branch, which is now at: https://github.com/CGAL/cgal-public-dev/tree/sdg-ptr-fix-philaris .

**Comment by sloriot** [comment]
I think this was fixed by #3362

**Comment by afabri** [comment]
I don't agree that the issue is fixed.   It makes no sense to ` typedef ....  ring_type`,   if `...`  must be that supports a division.
